common/ffsbuilder.cpp

```diff
@@ -379,34 +482,967 @@ USTATUS FfsBuilder::buildFreeSpace(const UModelIndex & index, UByteArray & freeS
 }
 
-USTATUS FfsBuilder::buildFile(const UModelIndex & index, UByteArray & file)
+USTATUS FfsBuilder::buildNvramStore(const UModelIndex & index, UByteArray & store)
 {
-    U_UNUSED_PARAMETER(index);
-    U_UNUSED_PARAMETER(file);
+    UByteArray header = model->header(index);
+    UByteArray body  = model->body(index);
+    UINT8 type = model->type(index);
+
+    if(model->action(index) == Actions::Remove) {
+        header.clear();
+        body.clear();
+    }
+    else if(model->action(index) == Actions::Replace ||
+            model->action(index) == Actions::Rebuild) {
+        if(type == Types::FdcStore) {
+            // Recalculate store header
+            EFI_FIRMWARE_VOLUME_HEADER* volumeHeader = (EFI_FIRMWARE_VOLUME_HEADER*)header.data();
+            volumeHeader->Checksum = 0;
+            volumeHeader->Checksum = calculateChecksum16((const UINT16*)volumeHeader, volumeHeader->HeaderLength);
+
+            // Rebuild VSS or VSS2 volume inside
+            UByteArray vssStore;
+            buildNvramStore(index.child(0, 0), vssStore);
+            body = vssStore;
+        }
+        else if(type == Types::VssStore || type == Types::Vss2Store) {
+            if(model->rowCount(index)) {
+                body.clear();
+                for (int i = 0; i < model->rowCount(index); i++) {
+                    UModelIndex currentIndex = index.child(i, 0);
+                    UByteArray currentHeader = model->header(currentIndex);
+                    UByteArray currentBody = model->body(currentIndex);
+                    UINT8 currentAction = model->action(currentIndex);
+
+                    if(currentAction == Actions::Remove) {
+                        currentHeader.clear();
+                        currentBody.clear();
+                    }
+                    else if(currentAction == Actions::Rebuild ||
+                            currentAction == Actions::Replace) {
+                        // Recalculate all Apple variables crc's
+                        if(model->subtype(currentIndex) == Subtypes::AppleVssEntry) {
+                            VSS_APPLE_VARIABLE_HEADER* appleVariableHeader = (VSS_APPLE_VARIABLE_HEADER*)currentHeader.data();
+                            appleVariableHeader->DataCrc32 = crc32(0, (const UINT8*)currentBody.constData(), currentBody.size());
+                            //appleVariableHeader->DataSize = currentBody.size();
+                        }
+                    }
+                    body += currentHeader + currentBody;
+                }
+            }
+        }
+        else if(type == Types::FsysStore) {
+             UByteArray store = header + body;
+
+            // Recalculate store checksum
+            UINT32 calculatedCrc = crc32(0, (const UINT8*)store.constData(), (const UINT32)store.size() - sizeof(UINT32));
+            // Write new checksum
+            body.replace((const UINT32)body.size() - sizeof(UINT32), sizeof(UINT32), (const char*)calculatedCrc, sizeof(UINT32));
```

Comment:
I think what was meant to be written here is not `(const char*)calculatedCrc` but `&calculatedCrc`. I will fix it up in a subsequent commit, but decided to let you know in case I misunderstood something on a higher level.